State gets corrupt when using attached disks

### Terraform Version
Terraform v0.10.8

### Affected Resource(s)
Please list the resources as a list, for example:
- google_compute_instance

### Terraform Configuration Files
```hcl
resource "google_compute_disk" "default" {
  name    = "boot-${var.project_name}"
  type    = "${var.pd_type}"
  zone    = "${var.zone}"
  image   = "${var.boot_image}"
  project = "${var.google_project}"
  size    = "${var.pd_size}"
}

resource "google_compute_disk" "files_pd" {
  name = "files-${var.project_name}"
  type = "${var.files_pd_type}"
  size = "${var.files_pd_size}"
  zone = "${var.zone}"

  project = "${var.google_project}"
}

resource "google_compute_instance" "default" {
  name         = "instance-${var.project_name}"
  machine_type = "${var.machine_type}"
  zone         = "${var.zone}"
  project      = "${var.google_project}"

  network_interface {
    subnetwork = "${var.google_subnetwork_name}"

    access_config {
      // Ephemeral IP
    }
  }

  boot_disk {
    source      = "${google_compute_disk.default.id}"
    auto_delete = false
  }

  attached_disk {
    source      = "${google_compute_disk.files_pd.id}"
    device_name = "files-disk"
  }
}
```

### Debug Output


### Panic Output


### Expected Behavior
The disk is always available and the state looks like this:

```
                "google_compute_instance.default": {
                    "type": "google_compute_instance",
                    "depends_on": [
                        "google_compute_disk.default",
                        "google_compute_disk.files_pd"
                    ],
                    "primary": {
                        "id": "project-name",
                        "attributes": {
                            "attached_disk.#": "1",
                            "attached_disk.0.device_name": "files-disk",
                            "attached_disk.0.disk_encryption_key_raw": "",
                            "attached_disk.0.disk_encryption_key_sha256": "",
                            "attached_disk.0.source": "https://www.googleapis.com/compute/v1/projects/project-name/zones/europe-west1-d/disks/files-project-name",
...
```

### Actual Behavior
The plan commands throws an error:
```
Error: Error refreshing state: 1 error(s) occurred:

* module.xxx.google_compute_instance.default: 1 error(s) occurred:

* module.xxx.google_compute_instance.default: google_compute_instance.default: The zonal field for resource disks cannot be empty.
```

Inspecting the state shows that it seems to be corrupted, the attached disk metadata are incomplete see below
```
                "google_compute_instance.default": {
                    "type": "google_compute_instance",
                    "depends_on": [
                        "google_compute_disk.default",
                        "google_compute_disk.files_pd",
                    ],
                    "primary": {
                        "id": "project-name",
                        "attributes": {
                            "attached_disk.#": "1",
....
```
### Steps to Reproduce
It's really happening randomly 😞  and so far couldn't figure much of what was triggering it ...

EDIT:
I was able to fix the state manually by completing the attached_disk metadata, plan then looks like:

```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  ~ module.xxx.google_compute_instance.default
      attached_disk.0.device_name: "" => "files-disk"
      attached_disk.0.source:      "" => "files-project-name"


Plan: 0 to add, 1 to change, 0 to destroy.
```

Applying this will corrupt the state again.

@danawillow I actually just had this happen to me today. I'm unable to run any terraform commands plan, apply or destroy against the resource.
```
:~$ terraform -version
Terraform v0.11.3
+ provider.google v1.6.0
```

```
TF_LOG=debug terraform plan -target google_compute_instance.worker[0]
```

```
...
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "disks": [
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   {
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "kind": "compute#attachedDisk",
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "type": "PERSISTENT",
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "mode": "READ_WRITE",
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "source": "https://www.googleapis.com/compute/v1/projects/i-test-dev/zones/us-central1-a/disks/dev-swarm-worker-1",
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "deviceName": "persistent-disk-0",
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "index": 0,
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "boot": true,
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "autoDelete": true,
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "licenses": [
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:     "https://www.googleapis.com/compute/v1/projects/ubuntu-os-cloud/global/licenses/ubuntu-1604-xenial"
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    ],
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "interface": "SCSI"
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   }2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  ],
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "metadata": {
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   "kind": "compute#metadata",
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   "fingerprint": "vZJWiOJtNqI="
2018-02-26T00:54:37.607-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  },
...
```

```
2018/02/26 00:54:37 [ERROR] root: eval: *terraform.EvalRefresh, err: google_compute_instance.worker.0: The zonal field for resource disks cannot be empty.
2018/02/26 00:54:37 [ERROR] root: eval: *terraform.EvalSequence, err: google_compute_instance.worker.0: The zonal field for resource disks cannot be empty.
```

Sure thing @danawillow.

Here is my worker.tf file. 

```resource "google_compute_instance" "worker" {
  name         = "${terraform.env}-${var.name}-worker-${count.index+1}"
  count        = "${var.workers_instance_count}"
  machine_type = "${var.workers_instance_type}"
  zone         = "${element(var.zones, count.index)}"

  lifecycle {
      ignore_changes = ["scratch_disk", "attached_disk"]
  }

  boot_disk {
    initialize_params{
      image = "${var.image}"
      size = "${var.disk_size}"
    }
  }

  network_interface {
    subnetwork = "${var.subnetwork}"
    subnetwork_project = "${var.subnetwork_project}"

    access_config {
      # ephemeral
    }
  }

  service_account {
    email = "${var.service_account_email}"
    scopes = "${var.service_account_scopes}"
  }

  tags = "${var.network_tags}"
}
```
Output from  `terraform plan -target google_compute_instance.worker[0]`
```
➜  swarm git:(master) ✗ terraform plan -target google_compute_instance.worker[0]
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

google_compute_instance.worker[0]: Refreshing state... (ID: dev-swarm-worker-1)
Releasing state lock. This may take a few moments...

Error: Error refreshing state: 1 error(s) occurred:

* google_compute_instance.worker: 1 error(s) occurred:

* google_compute_instance.worker[0]: google_compute_instance.worker.0: The zonal field for resource disks cannot be empty.
```

Debug output of the above

```
google_compute_instance.worker[0] - *terraform.NodeRefreshableManagedResourceInstance
google_compute_instance.worker[1] - *terraform.NodeRefreshableManagedResourceInstance
google_compute_instance.worker[2] - *terraform.NodeRefreshableManagedResourceInstance
2018/02/26 16:24:17 [DEBUG] Removing "google_compute_instance.worker[1]", filtered by targeting.
2018/02/26 16:24:17 [DEBUG] Removing "google_compute_instance.worker[2]", filtered by targeting.
2018/02/26 16:24:17 [DEBUG] ReferenceTransformer: "google_compute_instance.worker[0]" references: []
google_compute_instance.worker[0]: Refreshing state... (ID: dev-swarm-worker-1)
2018-02-26T16:24:17.824-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: 2018/02/26 16:24:17 [DEBUG] Google API Request Details:
2018-02-26T16:24:17.824-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: ---[ REQUEST ]---------------------------------------
2018-02-26T16:24:17.824-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: GET /compute/v1/projects/i-test-dev/zones/us-central1-a/instances/dev-swarm-worker-1?alt=json HTTP/1.1
2018-02-26T16:24:17.824-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: Host: www.googleapis.com
2018-02-26T16:24:17.824-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: User-Agent: google-api-go-client/0.5 (darwin amd64) Terraform/0.11.3-dev
2018-02-26T16:24:17.824-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: Accept-Encoding: gzip
2018-02-26T16:24:17.824-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:
2018-02-26T16:24:17.824-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:
2018-02-26T16:24:17.825-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: -----------------------------------------------------
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: 2018/02/26 16:24:18 [DEBUG] Google API Response Details:
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: ---[ RESPONSE ]--------------------------------------
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: HTTP/2.0 200 OK
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: Alt-Svc: hq=":443"; ma=2592000; quic=51303431; quic=51303339; quic=51303338; quic=51303337; quic=51303335,quic=":443"; ma=2592000; v="41,39,38,37,35"
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: Cache-Control: private, max-age=0, must-revalidate, no-transform
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: Content-Type: application/json; charset=UTF-8
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: Date: Mon, 26 Feb 2018 21:24:18 GMT
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: Etag: "cE7LX2cYo3Eyqx9AawYSQtO7CN4/fLGSMkull-IJLhETIgNozUaCZCI"
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: Expires: Mon, 26 Feb 2018 21:24:18 GMT
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: Server: GSE
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: Vary: Origin
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: Vary: X-Origin
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: X-Content-Type-Options: nosniff
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: X-Frame-Options: SAMEORIGIN
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: X-Xss-Protection: 1; mode=block
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: {
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "kind": "compute#instance",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "id": "4374421925458908157",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "creationTimestamp": "2017-11-09T20:37:39.391-08:00",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "name": "dev-swarm-worker-1",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "tags": {
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   "fingerprint": "42WmSpB8rSM="
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  },
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "machineType": "https://www.googleapis.com/compute/v1/projects/i-test-dev/zones/us-central1-a/machineTypes/n1-standard-8",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "status": "RUNNING",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "zone": "https://www.googleapis.com/compute/v1/projects/i-test-dev/zones/us-central1-a",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "networkInterfaces": [
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   {
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "kind": "compute#networkInterface",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "network": "https://www.googleapis.com/compute/v1/projects/i-test-dev/global/networks/company-network",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "subnetwork": "https://www.googleapis.com/compute/v1/projects/i-test-dev/regions/us-central1/subnetworks/i-test-dev-network",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "networkIP": "10.21.5.7",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "name": "nic0",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "accessConfigs": [
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:     {
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:      "kind": "compute#accessConfig",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:      "type": "ONE_TO_ONE_NAT",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:      "name": "external-nat",
2018-02-26T16:24:18.759-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:      "natIP": "35.248.25.12"
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:     }
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    ]
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   }
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  ],
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "disks": [
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   {
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "kind": "compute#attachedDisk",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "type": "PERSISTENT",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "mode": "READ_WRITE",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "source": "https://www.googleapis.com/compute/v1/projects/i-test-dev/zones/us-central1-a/disks/dev-swarm-worker-1",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "deviceName": "persistent-disk-0",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "index": 0,
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "boot": true,
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "autoDelete": true,
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "licenses": [
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:     "https://www.googleapis.com/compute/v1/projects/ubuntu-os-cloud/global/licenses/ubuntu-1604-xenial"
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    ],
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "interface": "SCSI"
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   }
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  ],
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "metadata": {
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   "kind": "compute#metadata",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   "fingerprint": "vZJWiOJtNqI="
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  },
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "serviceAccounts": [
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   {
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "email": "testdev@i-test-dev.iam.gserviceaccount.com",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    "scopes": [
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:     "https://www.googleapis.com/auth/devstorage.read_write",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:     "https://www.googleapis.com/auth/logging.write",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:     "https://www.googleapis.com/auth/compute"
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:    ]
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   }
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  ],
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "selfLink": "https://www.googleapis.com/compute/v1/projects/i-test-dev/zones/us-central1-a/instances/dev-swarm-worker-1",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "scheduling": {
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   "onHostMaintenance": "MIGRATE",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   "automaticRestart": true,
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:   "preemptible": false
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  },
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "cpuPlatform": "Intel Sandy Bridge",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "labelFingerprint": "42WmSpB8rSM=",
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "startRestricted": false,
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:  "deletionProtection": false
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: }
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4:
2018-02-26T16:24:18.760-0500 [DEBUG] plugin.terraform-provider-google_v1.6.0_x4: -----------------------------------------------------
2018/02/26 16:24:18 [ERROR] root: eval: *terraform.EvalRefresh, err: google_compute_instance.worker.0: The zonal field for resource disks cannot be empty.
2018/02/26 16:24:18 [ERROR] root: eval: *terraform.EvalSequence, err: google_compute_instance.worker.0: The zonal field for resource disks cannot be empty.
2018/02/26 16:24:18 [DEBUG] [aws-sdk-go] DEBUG: Request dynamodb/GetItem Details:
---[ REQUEST POST-SIGN ]-----------------------------
POST / HTTP/1.1
Host: dynamodb.us-east-1.amazonaws.com
User-Agent: aws-sdk-go/1.12.59 (go1.9.1; darwin; amd64) APN/1.0 HashiCorp/1.0 Terraform/0.11.3
Content-Length: 165
Accept-Encoding: identity
Content-Type: application/x-amz-json-1.0
X-Amz-Date: 20180226T212418Z
X-Amz-Target: DynamoDB_20120810.GetItem

{"ConsistentRead":true,"Key":{"LockID":{"S":"test-tf-state-dev-v2/env:/dev/swarm.tfstate"}},"ProjectionExpression":"LockID, Info","TableName":"test-tf-state-dev-v2"}
-----------------------------------------------------
2018/02/26 16:24:19 [DEBUG] [aws-sdk-go] DEBUG: Response dynamodb/GetItem Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Connection: close
Content-Length: 355
Content-Type: application/x-amz-json-1.0
Date: Mon, 26 Feb 2018 21:24:19 GMT
Server: Server


-----------------------------------------------------
2018/02/26 16:24:19 [DEBUG] [aws-sdk-go] {"Item":{"LockID":{"S":"test-tf-state-dev-v2/env:/dev/swarm.tfstate"},"Info":{"S":"{\"ID\":\"29157c68-1366-4562-3b48-b47c0a9a4abe\",\"Operation\":\"OperationTypePlan\",\"Info\":\"\",\"Who\":\"beenanner@beenanners-MacBook-Pro.local\",\"Version\":\"0.11.3\",\"Created\":\"2018-02-26T21:24:17.543152Z\",\"Path\":\"test-tf-state-dev-v2/env:/dev/swarm.tfstate\"}"}}}
2018/02/26 16:24:19 [DEBUG] [aws-sdk-go] DEBUG: Request dynamodb/DeleteItem Details:
---[ REQUEST POST-SIGN ]-----------------------------
POST / HTTP/1.1
Host: dynamodb.us-east-1.amazonaws.com
User-Agent: aws-sdk-go/1.12.59 (go1.9.1; darwin; amd64) APN/1.0 HashiCorp/1.0 Terraform/0.11.3
Content-Length: 105
Accept-Encoding: identity
Content-Type: application/x-amz-json-1.0
X-Amz-Date: 20180226T212419Z
X-Amz-Target: DynamoDB_20120810.DeleteItem

{"Key":{"LockID":{"S":"test-tf-state-dev-v2/env:/dev/swarm.tfstate"}},"TableName":"test-tf-state-dev-v2"}
-----------------------------------------------------
Releasing state lock. This may take a few moments...
2018/02/26 16:24:19 [DEBUG] [aws-sdk-go] DEBUG: Response dynamodb/DeleteItem Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Connection: close
Content-Length: 2
Content-Type: application/x-amz-json-1.0
Date: Mon, 26 Feb 2018 21:24:19 GMT
Server: Server


-----------------------------------------------------
2018/02/26 16:24:19 [DEBUG] [aws-sdk-go] {}
2018/02/26 16:24:19 [DEBUG] plugin: waiting for all plugin processes to complete...

Error: Error refreshing state: 1 error(s) occurred:

* google_compute_instance.worker: 1 error(s) occurred:

* google_compute_instance.worker[0]: google_compute_instance.worker.0: The zonal field for resource disks cannot be empty.


2018-02-26T16:24:19.388-0500 [DEBUG] plugin: plugin process exited: path=/Users/beenanner/company/docker/terraform/providers/gcp/us_central1/dev/swarm/.terraform/plugins/darwin_amd64/terraform-provider-google_v1.6.0_x4
```

--- a/google/resource_compute_instance.go
+++ b/google/resource_compute_instance.go
@@ -852,26 +852,26 @@ func resourceComputeInstanceRead(d *schema.ResourceData, meta interface{}) error
 		d.Set("label_fingerprint", instance.LabelFingerprint)
 	}
 
-	attachedDisksCount := d.Get("attached_disk.#").(int)
-	attachedDiskSources := make(map[string]int, attachedDisksCount)
-	for i := 0; i < attachedDisksCount; i++ {
-		source, err := ParseDiskFieldValue(d.Get(fmt.Sprintf("attached_disk.%d.source", i)).(string), d, config)
+	attachedDiskSources := make(map[string]int)
+	for i, v := range d.Get("attached_disk").([]interface{}) {
+		if v == nil {
+			continue
+		}
+		disk := v.(map[string]interface{})
+		source, err := ParseDiskFieldValue(disk["source"].(string), d, config)
 		if err != nil {
 			return err
 		}
 		attachedDiskSources[source.RelativeLink()] = i
 	}
 
-	sIndex := 0
-	attachedDisks := make([]map[string]interface{}, attachedDisksCount)
+	attachedDisks := make([]map[string]interface{}, d.Get("attached_disk.#").(int))
 	scratchDisks := []map[string]interface{}{}
-	extraAttachedDisks := []map[string]interface{}{}
 	for _, disk := range instance.Disks {
 		if disk.Boot {
 			d.Set("boot_disk", flattenBootDisk(d, disk, config))
 		} else if disk.Type == "SCRATCH" {
 			scratchDisks = append(scratchDisks, flattenScratchDisk(disk))
-			sIndex++
 		} else {
 			source, err := ParseDiskFieldValue(disk.Source, d, config)
 			if err != nil {
@@ -889,17 +889,26 @@ func resourceComputeInstanceRead(d *schema.ResourceData, meta interface{}) error
 				}
 				di["disk_encryption_key_sha256"] = key.Sha256
 			}
+			// We want the disks to remain in the order we set in the config, so if a disk
+			// is present in the config, make sure it's at the correct index. Otherwise, append it.
 			if inConfig {
 				attachedDisks[adIndex] = di
 			} else {
-				extraAttachedDisks = append(extraAttachedDisks, di)
+				attachedDisks = append(attachedDisks, di)
 			}
 		}
 	}
-	attachedDisks = append(attachedDisks, extraAttachedDisks...)
+	// Remove nils from map in case there were disks in the config that were not present on read;
+	// i.e. a disk was detached out of band
+	ads := []map[string]interface{}{}
+	for _, d := range attachedDisks {
+		if d != nil {
+			ads = append(ads, d)
+		}
+	}
 
 	d.Set("service_account", flattenServiceAccounts(instance.ServiceAccounts))
-	d.Set("attached_disk", attachedDisks)
+	d.Set("attached_disk", ads)
 	d.Set("scratch_disk", scratchDisks)
 	d.Set("scheduling", flattenScheduling(instance.Scheduling))
 	d.Set("guest_accelerator", flattenGuestAccelerators(instance.GuestAccelerators))

--- a/google/resource_compute_instance.go
+++ b/google/resource_compute_instance.go
@@ -855,6 +855,9 @@ func resourceComputeInstanceRead(d *schema.ResourceData, meta interface{}) error
 	attachedDiskSources := make(map[string]int)
 	for i, v := range d.Get("attached_disk").([]interface{}) {
 		if v == nil {
+			// There was previously a bug in this code that, when triggered,
+			// would cause some nil values to end up in the list of attached disks.
+			// Check for this case to make sure we don't try to parse the nil disk.
 			continue
 		}
 		disk := v.(map[string]interface{})

